### PR TITLE
Add bulkInsert and remove operations.

### PR DESCRIPTION
**Pull request by JoshDunnDev**

### CODE REVIEW

File: lib/restrictions.js [hunk 1]
```diff
@@ -73,6 +73,25 @@ export async function insert({restriction} = {}) {
   return record;
 }
 
+/**
+ * Inserts multiple restrictions into database.
+ *
+ * @param {object} options - The options to use.
+ * @param {Array} options.restrictions - An array of restrictions.
+ * @returns {Promise} - Settles once the operation completes.
```
Comment (dlongley) [inline lib/restrictions.js:81]:
```suggestion
 * @param {Array} options.restrictions - An array of restrictions.
 *
 * @returns {Promise} - Settles once the operation completes.
```

File: lib/restrictions.js [hunk 2]
```diff
@@ -73,6 +73,25 @@ export async function insert({restriction} = {}) {
...
+ */
+export async function bulkInsert({restrictions} = []) {
```
Comment (aljones15) [inline lib/restrictions.js:83]: I think this is a minor mistake here.

You are using object destructuring on an array when I believe you want:

```js
bulkInsert({restrictions = []} = {}) {
```
that will mean if no options are specified a blank object is provided,
then restrictions will default to an empty array.

File: lib/restrictions.js [hunk 3]
```diff
@@ -73,6 +73,25 @@ export async function insert({restriction} = {}) {
...
+  const operations = [];
+  for(const restriction of restrictions) {
+    const now = Date.now();
+    operations.push({
+      insertOne: {document: {restriction, meta: {created: now, updated: now}}}
```
Comment (aljones15) [inline lib/restrictions.js:88]: I'm assuming using bulkWrite here was intended, but if the only operation we are doing in bulk is `insertOne` then why not just use `insertMany`? Are we leaving ourselves open to doing multiple different operation types like `updateOne`, `deleteOne`, etc in a BulkWrite in the future?

File: lib/restrictions.js (hunk 1 above)
Comment (dmitrizagidulin) [inline lib/restrictions.js:81]:
```suggestion
 *
 * @returns {Promise} - Resolves once the operation completes.
```
(blank line before the `@returns` statement).

File: lib/restrictions.js [hunk 4]
```diff
@@ -73,6 +73,25 @@ export async function insert({restriction} = {}) {
...
+    });
+  }
+  const collection = database.collections['resource-restriction-restriction'];
+  await collection.bulkWrite(operations);
```
Comment (dlongley) [inline lib/restrictions.js:92]: Can we use `insertMany` instead? We can use the same date and meta for all of them as well.

```suggestion
  const now = Date.now();
  const meta = {created: now, updated: now};
  const records = restrictions.map(restriction => ({restriction, meta}));
  // allow unordered writes
  const writeOptions = {...database.writeOptions, ordered: false};
  const collection = database.collections['resource-restriction-restriction'];
  await collection.insertMany(records, writeOptions);
```

Not sure what we want to do with duplicates here, but for now I guess we're throwing. We may want to change that in the future.

**Review by dmitrizagidulin — APPROVED** [review]
👍 
Thanks @JoshDunnDev!

**Review by aljones15 — CHANGES_REQUESTED** [review]
Just one minor, totally understandable, and incredibly small destructuring issues keeps this from being perfecto. Great work BTW and thanks for the code. Feel free to ping me in future MongoDB related PRs.

File: lib/restrictions.js [hunk 5]
```diff
@@ -73,6 +73,25 @@ export async function insert({restriction} = {}) {
   return record;
 }
 
+/**
+ * Inserts multiple restrictions into database.
```
Comment (dlongley) [inline lib/restrictions.js:77]:
```suggestion
 * Inserts multiple restrictions into the database.
```

File: lib/restrictions.js [hunk 6]
```diff
@@ -136,7 +155,26 @@ export async function get({zone, resource} = {}) {
   return record;
 }
 
-// TODO: implement bulk insert, bulk update, and removal
+/**
+ * Deletes a restriction from database.
```
Comment (dlongley) [inline lib/restrictions.js:159]:
```suggestion
 * Deletes a restriction from the database.
```

File: lib/restrictions.js [hunk 7]
```diff
@@ -136,7 +155,26 @@ export async function get({zone, resource} = {}) {
...
+export async function remove({zone, resource} = {}) {
+  const query = {
+    'restriction.zone': zone,
+    'restriction.resource': resource
+  };
+  const collection = database.collections['resource-restriction-restriction'];
+  await collection.deleteOne(query);
+}
+
+// TODO: implement bulk update
```
Comment (dlongley) [inline lib/restrictions.js:177]: Remove and turn into an issue.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/restrictions.js (hunk 2 above)
Comment (dlongley) [reply lib/restrictions.js:83]: We should add `assert.array` here instead of setting a default, IMO.

**Review by dlongley — APPROVED** [review]

File: lib/restrictions.js [hunk 8]
```diff
@@ -5,6 +5,7 @@ import * as bedrock from 'bedrock';
 import * as database from 'bedrock-mongodb';
 import {promisify} from 'util';
 import {ResourceRestriction} from './ResourceRestriction.js';
+import assert from 'assert-plus';
```
Comment (dlongley) [inline lib/restrictions.js:8]: Please move this up to the top (alphabetize).

Comment (dmitrizagidulin) [reply lib/restrictions.js]: Fixed.

**Pull request merged**